modules/images/image-loading-optimization/class-ilo-url-metric.php

```diff
@@ -0,0 +1,181 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_URL_Metric class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Representation of the measurements taken from a single client's visit to a specific URL.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_URL_Metric implements JsonSerializable {
+
+	/**
+	 * Data.
+	 *
+	 * @var array{
+	 *          timestamp: int,
+	 *          viewport: array{ width: int, height: int },
+	 *          elements: array<array{
+	 *              isLCP: bool,
+	 *              isLCPCandidate: bool,
+	 *              xpath: string,
+	 *              intersectionRatio: float,
+	 *              intersectionRect: array{ width: int, height: int },
+	 *              boundingClientRect: array{ width: int, height: int },
+	 *          }>
+	 *      }
+	 */
+	private $data;
```

Comment:
Wouldn't it be valuable to store the `$url` and `$slug` in this class as well? Maybe they could be separate properties from `$data`, but I find it a bit odd to have a class for a URL metric exclude those identifying variables.

I think it could simplify some things by putting the responsibility for this information purely into the class representing the entity. We could still use associative arrays keyed by URL or slug if we need it, but I think it would be cleaner to have it covered by this class in its entirety - unless I'm missing something.

Alternatively, if there are scenarios where URL and slug should be decoupled from the actual _metric_ data, maybe we could separately introduce a class that is a value object representing URL and slug, and it could also provide the methods to create and validate the nonces etc. Just an idea.

Reply:
Well, the thing is that there are multiple URL metrics stored for a given response. The post type stores this list of URL metrics, based on the number of breakpoints and the sample size. So if we store the URL and slug with each `ILO_URL_Metric` we're storing multiple copies of the same data. The URL/slug is more higher level metadata which is only needed for the post type. We don't need the URL or slug except to get/set the post. I'd prefer to not include the URL or slug in this class.

Reply:
Well, let me think about this some more.

Reply:
Maybe we merge this as-is and I propose the change in another PR? I want to revisit how the URL is handled because in reality it is the request args really that matter, not the end URL. The slug is computed from the request args, not the URL.

Reply:
I've added this as a todo item on #869

Reply:
That makes sense. How about a separate class though to represent the URL + slug and related logic (e.g. `ILO_URL` or something like that)? Then an instance of that class could be passed to `ILO_URL_Metric`.

Indeed the URL is not _identifying_ data as it is used for several metrics, but it'll be very intuitive from an architecture perspective that via a `ILO_URL_Metric` instance you can find out which URL it is for. This is similar to any parent entity of another entity. From the child entity you can find out which the parent entity is.

Additionally, by using a class instance there's no duplication in terms of storage. You can pass the same `ILO_URL` instance to multiple `ILO_URL_Metric` instances, and since objects are passed by reference, they'd all reference the same data by definition. I think that would be intuitive, and it would unlock a few places where then you could pass only an `ILO_URL_Metric` instance. And then code that needs URL data could do something like `$metric->get_url()->get_url()`. Or maybe the class or at least the method could receive a different name to avoid this weirdness, such as `$metric->get_url_entity()->get_url()` which looks more intuitive.

Reply:
Ultimately, I think it would actually make sense to store a `url` with each URL Metric because in actuality the URL can vary while the request args change. This could be important for debugging, for example if it turns out that a template is modified by direct access to `$_GET` and not for a registered query var.

But I want to address this in a subsequent PR.

Reply:
Oh I see you added more comments in the meantime. Let me know what you think about the above proposal.

But in any case, this would be okay to do in a follow up PR.

Reply:
Yeah, let's do in a separate PR. It would actually be useful to duplicate the URL and even the request args for each URL metric since it reveals the underlying data for the response that created the URL metric. The `slug` used in the post type is computed from a normalized set of request args, so having the original request args would still be useful to have on hand.

Reply:
Sounds great!

Reply:
See https://github.com/WordPress/performance/pull/1043